MsCorePkg/Crates/RustBootServicesAllocatorDxe/src/lib.rs

```diff
@@ -153,31 +138,57 @@ impl BootServicesAllocator {
 /// Provides a locked allocator instance (using [`spin::Mutex`]) that is suitable for use as a
 /// [`core::alloc::GlobalAlloc`].
 pub struct SpinLockedAllocator {
+  boot_services: AtomicPtr<efi::BootServices>,
   inner: spin::Mutex<BootServicesAllocator>,
 }
 
 impl SpinLockedAllocator {
   // Create a new instance. const fn to allow static initialization.
   const fn new() -> Self {
-    SpinLockedAllocator { inner: spin::Mutex::new(BootServicesAllocator::new()) }
+    SpinLockedAllocator {
+      boot_services: AtomicPtr::new(core::ptr::null_mut()),
+      inner: spin::Mutex::new(BootServicesAllocator::new()),
+    }
   }
 
   /// Initialize the allocator.
   ///
   /// This routine initializes the allocator by providing a pointer to the global EFI Boot Services table that will
   /// supply pointers to the AllocatePool() and FreePool() primitives that implement memory allocation.
-  pub fn init(&self, boot_services: *mut BootServices) {
-    self.inner.lock().init(boot_services);
+  pub fn init(&self, boot_services: *mut efi::BootServices) {
+    self.boot_services.store(boot_services, core::sync::atomic::Ordering::SeqCst);
   }
 }
 
 unsafe impl GlobalAlloc for SpinLockedAllocator {
   unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-    self.inner.lock().boot_services_alloc(layout)
+    let bs_ptr = self.boot_services.load(core::sync::atomic::Ordering::SeqCst);
+    if let Some(boot_services) = unsafe { bs_ptr.as_ref() } {
+      let ptr;
+      let old_tpl = (boot_services.raise_tpl)(efi::TPL_NOTIFY);
+      if let Some(allocator) = self.inner.try_lock() {
+        ptr = allocator.boot_services_alloc(layout, boot_services);
+      } else {
+        panic!("failed to acquire memory allocation lock at TPL_NOTIFY");

```

Comment:
My understanding is that most of the panic handler implementations are currently going to print a `DEBUG_ERROR` message and then loop indefinitely. Is this expected to potentially occur in a production integration scenario? If so, is that the behavior you're intending for this case?

Reply:
I could alternatively silently return a null pointer to indicate allocation failed. However, gBS->Allocate*() is going to ASSERT with a TPL violation in the same condition, so I was thinking that panic would be a better aid to immediately identifying bad code. Happy to change it though, I actually implemented the null pointer return first, but then decided I was more opinionated about it - but I'm not _that_ opinionated.

Reply:
I'm not super opinionated either and I agree with bringing attention to unanticipated edge cases, but lock contention is something that is hard to predict, as it's very dynamic based on the flows involved on a per-platform basis.

My main concern is about potentially placing the system in an unrecoverable state if one of those instances that occurs more often than expected. Typically, we use the same assert behavior across our platforms which is similar to the panic handler implementations. But I thought the panic might apply that behavior (i.e. in `RELEASE` builds) diverging from our normal assert behavior in those cases.

That's why I'm leaning a little more in favor of relying upon the asserts in `gBS->Allocate()` for now.

Reply:
I think just removing the lock implementation altogether to defer to gBS->Allocate*() TPL management is probably the right way to go. Will update this PR with that change.

Reply:
Implementation updated; ready for a re-review.

Reply:
Thanks for pushing the update.